src/Client.php

```diff
@@ -40,19 +46,140 @@ public function sendRequest(RequestInterface $request)
                 /** @var Artax\Response $resp */
                 $resp = yield $this->client->request($req, [
                     Artax\Client::OP_MAX_REDIRECTS => 0,
-                ]);
+                ], $cancellationTokenSource->getToken());
             } catch (Artax\HttpException $e) {
                 throw new RequestException($e->getMessage(), $request, $e);
             }
 
             $respBody = $resp->getBody();
-            $bodyStream = $this->streamFactory->createStream();
+            $bodyStream = new class($respBody, $cancellationTokenSource) implements StreamInterface {
+                private $buffer = '';
+                private $position = 0;
+                private $eof = false;
 
-            while (null !== $chunk = yield $respBody->read()) {
-                $bodyStream->write($chunk);
-            }
+                private $body;
+                private $cancellationTokenSource;
+
+                public function __construct(Message $body, CancellationTokenSource $cancellationTokenSource)
+                {
+                    $this->body = $body;
+                    $this->cancellationTokenSource = $cancellationTokenSource;
+                }
+
+                public function __toString()
+                {
+                    try {
+                        return $this->getContents();
+                    } catch (\Throwable $e) {
+                        return '';
+                    }
+                }
+
+                public function __destruct()
+                {
+                    $this->close();
+                }
+
+                public function close()
+                {
+                    $this->cancellationTokenSource->cancel();
```

Comment:
not a pro on amp, it does not give problems if the stream is already closed to cancel it twice ?